File: packages/py/analytics/src/bolsa_analytics/cognitive/operating_policy.py
"""OperatingPolicy — composición InvestorProfile → políticas operativas (V1.35)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from bolsa_analytics.cognitive.exit_policy import ExitPolicy, resolve_exit_policy
from bolsa_analytics.cognitive.trading_policy import TradingPolicy

OperatingPolicyTemplateId = Literal["conservative", "moderate", "aggressive_swing"]


@dataclass(frozen=True, slots=True)
class EntryPolicy:
    requires_setup: bool = True


@dataclass(frozen=True, slots=True)
class PositionSizingPolicy:
    max_open_positions: int
    max_portfolio_concentration_pct: float


@dataclass(frozen=True, slots=True)
class TrailingPolicy:
    trail_width: str
    ratchet_only: bool = True


@dataclass(frozen=True, slots=True)
class TimePolicy:
    max_holding_period_days: int
    min_holding_period_minutes: int


@dataclass(frozen=True, slots=True)
class ConcentrationPolicy:
    max_sector_exposure_pct: float
    max_leverage: float


@dataclass(frozen=True, slots=True)
class OperatingPolicy:
    template_id: OperatingPolicyTemplateId
    entry: EntryPolicy
    risk: Any
    sizing: PositionSizingPolicy
    exit: ExitPolicy
    trailing: TrailingPolicy
    time: TimePolicy
    concentration: ConcentrationPolicy

    def to_dict(self) -> dict[str, Any]:
        return {
            "templateId": self.template_id,
            "entry": {"requiresSetup": self.entry.requires_setup},
            "risk": dict(self.risk.to_dict()) if hasattr(self.risk, "to_dict") else {},
            "sizing": {
                "maxOpenPositions": self.sizing.max_open_positions,
                "maxPortfolioConcentrationPct": self.sizing.max_portfolio_concentration_pct,
            },
            "exit": self.exit.to_dict(),
            "trailing": {
                "trailWidth": self.trailing.trail_width,
                "ratchetOnly": self.trailing.ratchet_only,
            },
            "time": {
                "maxHoldingPeriodDays": self.time.max_holding_period_days,
                "minHoldingPeriodMinutes": self.time.min_holding_period_minutes,
            },
            "concentration": {
                "maxSectorExposurePct": self.concentration.max_sector_exposure_pct,
                "maxLeverage": self.concentration.max_leverage,
            },
        }


def _normalize_template(template_id: str | None) -> OperatingPolicyTemplateId:
    if template_id == "conservative":
        return "conservative"
    if template_id == "aggressive_swing":
        return "aggressive_swing"
    return "moderate"


def _resolve_trading_policy(template_id: OperatingPolicyTemplateId) -> TradingPolicy:
    from bolsa_analytics.cognitive.trading_policy_templates import get_policy_template

    return get_policy_template(template_id)


def resolve_operating_policy(template_id: str | None = None) -> OperatingPolicy:
    tid = _normalize_template(template_id)
    trading = _resolve_trading_policy(tid)
    exit_policy = resolve_exit_policy(tid)
    return OperatingPolicy(
        template_id=tid,
        entry=EntryPolicy(),
        risk=trading.risk,
        sizing=PositionSizingPolicy(
            max_open_positions=trading.exposure.max_open_positions,
            max_portfolio_concentration_pct=trading.exposure.max_portfolio_concentration_pct,
        ),
        exit=exit_policy,
        trailing=TrailingPolicy(trail_width=exit_policy.trail_width),
        time=TimePolicy(
            max_holding_period_days=trading.horizon.max_holding_period_days,
            min_holding_period_minutes=trading.horizon.min_holding_period_minutes,
        ),
        concentration=ConcentrationPolicy(
            max_sector_exposure_pct=trading.exposure.max_sector_exposure_pct,
            max_leverage=trading.exposure.max_leverage,
        ),
    )

File: packages/shared/src/cognitive/operating-policy.ts
/**
 * OperatingPolicy — composición InvestorProfile → políticas operativas (V1.35).
 * Esqueleto documental: compone TradingPolicy + ExitPolicy sin nuevos gates.
 */

import type { ExitPolicyV1 } from "./exit-policy.js";
import { resolveExitPolicy } from "./exit-policy.js";
import type { PolicyTemplateId, TradingPolicyV1 } from "./trading-policy.js";
import { getTradingPolicyTemplate } from "./trading-policy-templates.js";

export type OperatingPolicyTemplateId = Extract<
  PolicyTemplateId,
  "conservative" | "moderate" | "aggressive_swing"
>;

export type EntryPolicyV1 = {
  /** Placeholder — setup bias / entry_ready (TradePlan v0). */
  requiresSetup: boolean;
};

export type RiskPolicyV1 = TradingPolicyV1["risk"];

export type PositionSizingPolicyV1 = {
  maxOpenPositions: number;
  maxPortfolioConcentrationPct: number;
};

export type TrailingPolicyV1 = {
  trailWidth: ExitPolicyV1["trailWidth"];
  /** Ratchet: nunca empeorar stop sin override auditado. */
  ratchetOnly: true;
};

export type TimePolicyV1 = {
  maxHoldingPeriodDays: number;
  minHoldingPeriodMinutes: number;
};

export type ConcentrationPolicyV1 = {
  maxSectorExposurePct: number;
  maxLeverage: number;
};

export type OperatingPolicyV1 = {
  templateId: OperatingPolicyTemplateId;
  entry: EntryPolicyV1;
  risk: RiskPolicyV1;
  sizing: PositionSizingPolicyV1;
  exit: ExitPolicyV1;
  trailing: TrailingPolicyV1;
  time: TimePolicyV1;
  concentration: ConcentrationPolicyV1;
};

function normalizeTemplateId(
  templateId: string | null | undefined,
): OperatingPolicyTemplateId {
  if (templateId === "conservative" || templateId === "aggressive_swing") {
    return templateId;
  }
  return "moderate";
}

/** InvestorProfile template → OperatingPolicy (≠ permiso de firma). */
export function resolveOperatingPolicy(
  templateId: string | null | undefined,
): OperatingPolicyV1 {
  const id = normalizeTemplateId(templateId);
  const trading = getTradingPolicyTemplate(id);
  const exit = resolveExitPolicy(id);
  return {
    templateId: id,
    entry: { requiresSetup: true },
    risk: trading.risk,
    sizing: {
      maxOpenPositions: trading.exposure.maxOpenPositions,
      maxPortfolioConcentrationPct:
        trading.exposure.maxPortfolioConcentrationPct,
    },
    exit,
    trailing: {
      trailWidth: exit.trailWidth,
      ratchetOnly: true,
    },
    time: {
      maxHoldingPeriodDays: trading.horizon.maxHoldingPeriodDays,
      minHoldingPeriodMinutes: trading.horizon.minHoldingPeriodMinutes,
    },
    concentration: {
      maxSectorExposurePct: trading.exposure.maxSectorExposurePct,
      maxLeverage: trading.exposure.maxLeverage,
    },
  };
}

export function formatOperatingPolicyPreview(
  policy: OperatingPolicyV1,
): string {
  const t1 = Math.round(policy.exit.t1ReduceFraction * 100);
  const t2 = Math.round(policy.exit.t2ReduceFraction * 100);
  return (
    `${policy.templateId}: T1 ${t1}% · T2 ${t2}% · trail ${policy.trailing.trailWidth} · ` +
    `max DD ${policy.risk.hardMaxDrawdownLimitPct}% · sector ${policy.concentration.maxSectorExposurePct}%`
  );
}
